src/sdk/namespace_fs.js

```diff
@@ -682,44 +682,83 @@ class NamespaceFS {
-        const upload_id = uuidv4();
-        const upload_path = path.join(this.bucket_path, this.get_bucket_tmpdir(), 'uploads', upload_id);
-        let fs_xattr = to_fs_xattr(params.xattr);
-        // dbg.log0('NamespaceFS.upload_object:', upload_path, '->', file_path);
         try {
-            await this._check_path_in_bucket_boundaries(fs_account_config, file_path);
-            await this._make_path_dirs(upload_path, fs_account_config);
-            if (params.copy_source) {
-                const source_file_path = path.join(this.bucket_path, params.copy_source.key);
-                await this._check_path_in_bucket_boundaries(fs_account_config, source_file_path);
-                await this._fail_if_archived_or_sparse_file(fs_account_config, source_file_path);
-                try {
-                    const same_inode = await this._is_same_inode(fs_account_config, source_file_path, file_path, fs_xattr);
-                    if (same_inode) return same_inode;
-                    // Doing a hard link.
-                    await nb_native().fs.link(fs_account_config, source_file_path, upload_path);
-                } catch (e) {
-                    dbg.warn('NamespaceFS: COPY using link failed with:', e);
-                    fs_xattr = await this._copy_stream(source_file_path, upload_path, fs_account_config, fs_xattr);
+            if (can_use_temp_file) {
+                const open_mode = 'wt';
+                // Uploading to a temp file then we will rename it.
+                let fs_xattr = to_fs_xattr(params.xattr);
+                // dbg.log0('NamespaceFS.upload_object:', this.upload_path, '->', file_path);
+                if (params.copy_source) {
+                    const upload_id = uuidv4();
+                    const upload_path = path.join(this.bucket_path, this.get_bucket_tmpdir(), 'uploads', upload_id);
+                    const source_file_path = path.join(this.bucket_path, params.copy_source.key);
+                    await this._check_path_in_bucket_boundaries(fs_account_config, file_path);
+                    await this._make_path_dirs(upload_path, fs_account_config);
+                    await this._check_path_in_bucket_boundaries(fs_account_config, source_file_path);
+                    await this._fail_if_archived_or_sparse_file(fs_account_config, source_file_path);
+                    try {
+                        const same_inode = await this._is_same_inode(fs_account_config, source_file_path, file_path, fs_xattr);
+                        if (same_inode) return same_inode;
+                        // Doing a hard link.
+                        await nb_native().fs.link(fs_account_config, source_file_path, upload_path);
+                    } catch (e) {
+                        dbg.warn('NamespaceFS: COPY using link failed with:', e);
+                        fs_xattr = await this._copy_stream(source_file_path, upload_path, fs_account_config, fs_xattr);
+                    }
```

Comment:
the remaining if(copy_source) block is exactly the same, so we should push it out of `if (can_use_temp_file)` to write it once in this flow.